### PR TITLE
FermiNet Training Complete (Bacward pass + Forward)

### PR DESCRIPTION
**Pull request by shaipranesh2**
## Description

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/tests/test_ferminet.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ def test_evaluate_hf_solution():
 
 
 @pytest.mark.dqc
-def test_FerminetMode_pretrain():
+def test_FerminetModel_pretrain():
```
Comment (shaipranesh2) [inline deepchem/models/tests/test_ferminet.py:55]: Chaning the typo from previous PRs

File: deepchem/models/torch_models/ferminet.py [hunk 2]
```diff
@@ -164,12 +164,19 @@ def loss(self,
...
         """
         criterion = torch.nn.MSELoss()
-        if pretrain:
+        if pretrain[0]:
             psi_up_mo_torch = torch.from_numpy(psi_up_mo).unsqueeze(1)
             psi_down_mo_torch = torch.from_numpy(psi_down_mo).unsqueeze(1)
-            self.running_diff = self.running_diff + criterion(
-                self.psi_up, psi_up_mo_torch.float()) + criterion(
-                    self.psi_down, psi_down_mo_torch.float())
+            self.running_diff = self.running_diff.float() + criterion(
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py:170]: making it float type as psi)up_mo_torch is double and gives errors when run.
Also added loss for training in the else part.

File: deepchem/models/torch_models/ferminet.py [hunk 3]
```diff
@@ -476,42 +484,151 @@ def random_walk(self, x: np.ndarray) -> np.ndarray:
         A numpy array containing the joint probability of the hartree fock and the sampled electron's position coordinates
         """
         x_torch = torch.from_numpy(x).view(self.batch_no, -1, 3)
-        x_torch.requires_grad = True
+        if self.tasks == 'pretraining':
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py:487]: adding pretraining and training part separately in the  random_Walk fucntion

File: deepchem/models/torch_models/ferminet.py [hunk 4]
```diff
@@ -476,42 +484,151 @@ def random_walk(self, x: np.ndarray) -> np.ndarray:
...
+                        std_init *= 1.1
+                    else:
+                        std_init /= 1.1
                 self.loss_value = (torch.mean(self.model.running_diff) /
                                    self.random_walk_steps)
                 self.loss_value.backward()
                 optimizer.step()
                 self.model.running_diff = torch.zeros(self.batch_no)
+
+        if (self.tasks == 'training'):
```
Comment (shaipranesh2) [inline deepchem/models/torch_models/ferminet.py:587]: adding the training part

File: deepchem/models/torch_models/ferminet.py [hunk 5]
```diff
@@ -164,12 +164,19 @@ def loss(self,
             indicates whether the model is pretraining
```
Comment (rbharath) [inline deepchem/models/torch_models/ferminet.py:164]: A List[bool] is very awkward; let's swap this to True/False in a subsequent PR

File: deepchem/models/torch_models/ferminet.py [hunk 6]
```diff
@@ -476,42 +484,151 @@ def random_walk(self, x: np.ndarray) -> np.ndarray:
...
               nb_epoch: int = 200,
-              lr: float = 0.0075,
-              weight_decay: float = 0.0001):
+              lr: float = 0.002,
+              weight_decay: float = 0,
+              std: float = 0.08,
+              std_init: float = 0.02,
+              steps_std: int = 100):
         """
         function to run training or pretraining.
```
Comment (rbharath) [inline deepchem/models/torch_models/ferminet.py:539]: This docstring should explain in detail why we need to overwrite the TorchModel implementation

File: deepchem/models/torch_models/ferminet.py [hunk 7]
```diff
@@ -462,7 +470,7 @@ def prepare_hf_solution(self):
         self.mf = pyscf.scf.UHF(self.mol)
         _ = self.mf.kernel()
 
-    def random_walk(self, x: np.ndarray) -> np.ndarray:
+    def random_walk(self, x: np.ndarray):
         """
         Function to be passed on to electron sampler for random walk and gets called at each step of sampling
```
Comment (rbharath) [inline deepchem/models/torch_models/ferminet.py:476]: Document why the random walk is different for pretraining and finetuning

File: deepchem/models/torch_models/ferminet.py [hunk 8]
```diff
@@ -476,42 +484,151 @@ def random_walk(self, x: np.ndarray) -> np.ndarray:
...
+
+        if self.tasks == 'training':
+            energy = self.model.loss(pretrain=[False])
+            self.energy_sampled: torch.Tensor = torch.cat(
+                (self.energy_sampled, energy.unsqueeze(0)))
+            return 2 * np.log(np.abs(np_output))
+
+    def prepare_train(self, burn_in: int = 100):
+        """
+        Function to perform burn-in and to change the model parameters for training.
```
Comment (rbharath) [inline deepchem/models/torch_models/ferminet.py:515]: More details on why this is necessary

File: deepchem/models/torch_models/ferminet.py (hunk 6 above)
Comment (rbharath) [reply deepchem/models/torch_models/ferminet.py:539]: Explain in detail how pretrain works vs train works. Multiple paragraphs please

File: deepchem/models/torch_models/ferminet.py [hunk 9]
```diff
@@ -476,42 +484,151 @@ def random_walk(self, x: np.ndarray) -> np.ndarray:
...
+
         if (self.tasks == 'pretraining'):
-            for _ in range(nb_epoch):
+            for iteration in range(nb_epoch):
                 optimizer.zero_grad()
-                self.molecule.move()
+                accept = self.molecule.move(stddev=std_init)
+                if iteration % steps_std == 0:
+                    if accept > 0.55:
+                        std_init *= 1.1
```
Comment (rbharath) [inline deepchem/models/torch_models/ferminet.py:578]: Magic numbers are bad; These need to be documented or tunable parameters

**Review by rbharath — APPROVED** [review]
LGTM

@shaipranesh2 Please fix the documentation requests in a follow up PR
